### PR TITLE
Support inner doc attributes with `#!` & `//!` syntax

### LINKED ISSUE #3171: Support nested module attributes `#![attr]` - i.e. attributes intended for the enclosing module
Not sure if this is the right name for them, but this will likely be necessary for module-level documentation. E.g.

```rust
//! These are some module docs talking about `foo`.

library foo;

// ...
```

The `//!` doc comment at the top should convert into an attribute that looks something like:
```rust
#![doc("These are some module docs talking about `foo`.")]
```
which would be added to the set of attributes associated with the module `foo`.

### PR DESCRIPTION
**Pull request by eureka-cpu**
## Description
This adds the `HashBangToken` (`#!`) which will support nested module doc comments as well as inner doc comments for annotated items.

- [x] Add `HashBangToken`
- [x] Differentiate hash tokens on `AttributeDecl`
- [ ] Fix directional attribute parsing for `Inner` & `Outer` doc comments ( This is optional, the only downside of not implementing this right away is that inner and outer docs are treated equally )
- [x] Add attributes to `Module`
- [x] Complete support for module level attributes

## Checklist

- [x] I have linked to any relevant issues. Closes #3171 
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant. @mohammadfawaz 
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by eureka-cpu** [comment]
I made a test branch to try out the module level docs and........


https://user-images.githubusercontent.com/57543709/218918905-ebe1b46f-1d14-4cf0-b700-5b47d5cb0c46.mov

File: sway-parse/src/item/mod.rs [hunk 1]
```diff
@@ -206,9 +206,9 @@ mod tests {
         let item = parse_item(
             r#"
             // I will be ignored.
-            //! I will be ignored.
+            //! I can't feel the way I did before
```
Comment (mitchmindtree) [inline sway-parse/src/item/mod.rs:209]: :smiling_face_with_tear:

File: sway-parse/src/module.rs [hunk 2]
```diff
@@ -23,16 +30,50 @@ impl Parse for ModuleKind {
...
+            // TODO: Use a Literal instead of an Ident when Attribute args
+            // start supporting them and remove `Ident::new_no_trim`.
+            let value = Ident::new_no_trim(doc_comment.content_span.clone());
+            let hash_kind = match &doc_comment.doc_style {
+                DocStyle::Inner => Ok(AttributeHashKind::Inner(HashBangToken::new(
+                    doc_comment.span.clone(),
+                ))),
+                DocStyle::Outer => Err(parser.emit_error(ParseErrorKind::ExpectedModuleDocComment)),
+            }
+            .unwrap();
```
Comment (mitchmindtree) [inline sway-parse/src/module.rs:48]: Do we want to `unwrap` here? Or should we return an error somehow? (Not exactly sure how `ParseResult` works off the top of my head).

**Review by mitchmindtree — COMMENTED** [review]
Awesome stuff @eureka-cpu! Very exciting, I think this will do wonders for the accessibility of our docs. That `std` root module is begging for a welcome msg :smiling_face_with_tear: Just a couple tiny suggestions, otherwise looks great.

File: sway-parse/src/item/mod.rs (hunk 1 above)
Comment (eureka-cpu) [reply sway-parse/src/item/mod.rs:209]: https://youtu.be/LYU-8IFcDPw 🤘🏻

**Comment by eureka-cpu** [comment]
> Awesome stuff @eureka-cpu! Very exciting, I think this will do wonders for the accessibility of our docs. That `std` root module is begging for a welcome msg 🥲 Just a couple tiny suggestions, otherwise looks great.

I'll open an issue for it but if we want to have the project root module to have a description I think we should add a description option in the project section of the `forc.toml`. What do you think?

File: sway-parse/src/module.rs (hunk 2 above)
Comment (eureka-cpu) [reply sway-parse/src/module.rs:48]: I'll look into it. I was using `Result` in the heat of the moment as a wrapper type.

File: sway-core/src/lib.rs [hunk 3]
```diff
@@ -100,6 +103,53 @@ pub fn parse_tree_type(input: Arc<str>) -> CompileResult<parsed::TreeType> {
...
+                _ => None,
+            } {
+                match attrs_map.get_mut(&attr_kind) {
+                    Some(old_args) => {
+                        old_args.push(attribute);
+                    }
+                    None => {
+                        attrs_map.insert(attr_kind, vec![attribute]);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply sway-core/src/lib.rs:146]: ah nice, thanks (:

**Comment by mitchmindtree** [comment]
> I'll open an issue for it but if we want to have the project root module to have a description I think we should add a description option in the project section of the forc.toml. I could then use that description as the project index attribute. What do you think?

Fwiw, this would differ a little to how cargo and Rust work today. E.g.  `description`s in `Cargo.toml` files tend to be quite short and concise, and crates.io favours rendering the project's README on the project page. docs.rs on the other hand (closer to our target) first presents the crate-root-level documentation (i.e. the root module documentation, so `main.rs` or `lib.rs`). Often these docs can be large and detailed and give an overview of how to navigate the package ([std](https://doc.rust-lang.org/stable/std/) is a good example). I think modelling this behaviour probably makes the most sense.

Oh btw, I think it's totally fine that we add std root docs in a follow-up PR! Just getting excited about it as all :)

**Review by mitchmindtree — APPROVED** [review]
utACK

**Review by otrho — APPROVED** [review]

**Pull request merged**